// learning-platform-shared/schema.js
// Simple schema definitions for in-memory storage
// These are used for validation and type checking

export const userSchema = {
  id: "string",
  username: "string",
  email: "string", 
  password: "string",
  fullName: "string",
  role: "string", // student, instructor
  profileImage: "string",
  bio: "string",
  skills: "array", // Array of skill objects
  createdAt: "date",
};

export const learningPathSchema = {
  id: "string",
  userId: "string",
  title: "string",
  description: "string",
  modules: "array", // Array of module objects
  difficulty: "string", // beginner, intermediate, advanced
  estimatedHours: "number",
  isActive: "boolean",
  createdAt: "date",
};

export const userProgressSchema = {
  id: "string",
  userId: "string",
  learningPathId: "string",
  moduleId: "string",
  status: "string", // not_started, in_progress, completed
  progressPercentage: "number",
  hoursSpent: "number",
  lastAccessed: "date",
  updatedAt: "date",
};

export const achievementSchema = {
  id: "string",
  userId: "string",
  title: "string",
  description: "string",
  badgeIcon: "string",
  points: "number",
  unlockedAt: "date",
};

export const forumPostSchema = {
  id: "string",
  userId: "string",
  title: "string",
  content: "string",
  category: "string",
  tags: "array",
  likes: "number",
  replies: "number",
  isResolved: "boolean",
  createdAt: "date",
};

export const forumReplySchema = {
  id: "string",
  userId: "string",
  postId: "string",
  content: "string",
  likes: "number",
  isAccepted: "boolean",
  createdAt: "date",
};

export const opportunitySchema = {
  id: "string",
  title: "string",
  description: "string",
  type: "string", // internship, hackathon, workshop, job
  company: "string",
  location: "string",
  isRemote: "boolean",
  requirements: "array",
  applicationUrl: "string",
  deadline: "date",
  createdAt: "date",
};

export const resourceSchema = {
  id: "string",
  title: "string",
  description: "string",
  type: "string", // video, article, course, tutorial
  platform: "string", // youtube, coursera, udemy, geeksforgeeks
  url: "string",
  thumbnail: "string",
  duration: "string",
  rating: "number",
  difficulty: "string", // beginner, intermediate, advanced
  tags: "array",
  isFree: "boolean",
  price: "string",
  createdAt: "date",
};

// Simple validation functions
export const insertUserSchema = (data) => ({ ...data, id: data.id || Date.now().toString() });
export const insertLearningPathSchema = (data) => ({ ...data, id: data.id || Date.now().toString() });
export const insertUserProgressSchema = (data) => ({ ...data, id: data.id || Date.now().toString() });
export const insertAchievementSchema = (data) => ({ ...data, id: data.id || Date.now().toString() });
export const insertForumPostSchema = (data) => ({ ...data, id: data.id || Date.now().toString() });
export const insertForumReplySchema = (data) => ({ ...data, id: data.id || Date.now().toString() });
export const insertOpportunitySchema = (data) => ({ ...data, id: data.id || Date.now().toString() });
export const insertResourceSchema = (data) => ({ ...data, id: data.id || Date.now().toString() });
